record_processor.py

```diff
@@ -0,0 +1,65 @@
+import base64
+import csv
+import io
+import json
+
+from nypl_py_utils.functions.log_helper import create_log
+from nypl_py_utils.classes.avro_client import AvroDecoder
+
+
+class RecordProcessor:
+    def __init__(self, schema_url):
+        self.logger = create_log("record_processor")
+        self.avro_decoder = AvroDecoder(schema_url)
+        self.schema = self.avro_decoder.get_json_schema(schema_url)
+
+    def process_record(self, record, output_format):
+        """
+        Maps records to decoded Avro. Method returns data in
+        desired output format for Firehose (JSON or CSV string)
+        """
+        binary_record_data = base64.b64decode(record["data"])
+        decoded_record_data = self.avro_decoder.decode_record(binary_record_data)
+
+        if decoded_record_data is None:
+            # Unable to decode Avro record
+            return {
+                "recordId": record["recordId"],
+                "result": "ProcessingFailed",
+                "data": record["data"],
+            }
+        else:
+            result_string = self._format_result_string(
+                output_format, decoded_record_data
+            )
+            return {
+                "recordId": record["recordId"],
+                "result": "Ok",
+                "data": result_string,  # needed for JSON conversion
+            }
+
+    def _format_result_string(self, output_format, data):
+        if output_format == "csv" and isinstance(data, dict):
+            data = self._transform_dictionary_to_csv_string(data)
+        else:
+            data = json.dumps(data)
+        # After decoding, convert to base64. More often than not,
+        # the original data is either encoded or not in base64
+        to_bytes = data.encode("utf-8")
+        return (base64.b64encode(to_bytes)).decode("utf-8")
```

Comment:
Sorry I don't totally get this -- could explain a little more what's going on here?

Reply:
This is a little convoluted but we need to encode the decoded data to base64, and `base64.b64encode()` requires a bytes-like input. In addition, the `.decode("utf-8")` removes the byte wrapper around the decoded output. I.e. transforms `b'Y2N8TGlicmFyeSBcfCBDICEsPXxXZWR8MDE6MDB8MjM6MDB8MjAyMy0wMy0wMwo='` to `Y2N8TGlicmFyeSBcfCBDICEsPXxXZWR8MDE6MDB8MjM6MDB8MjAyMy0wMy0wMwo=`.

Reply:
So `data` is a regular string, then you convert it to UTF-8 bytes, then you convert those UTF-8 bytes to hex bytes, then you convert those hex bytes to a hex string? Could you add a comment saying that `data` is a regular string and we want to output that same string in hex, which requires using bytes as an intermediate type?

Reply:
Yes exactly. I'll add the comment in rn!